server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskList.java

```diff
@@ -0,0 +1,122 @@
+    // root.next points to the head
+    // root.prev points to the tail
+    private TimerTaskEntry root;
+
+    TimerTaskList(
+        AtomicInteger taskCounter
+    ) {
+        this.taskCounter = taskCounter;
+        this.expiration = new AtomicLong(-1L);
+        this.root = new TimerTaskEntry(null, -1L);
+        this.root.next = root;
+        this.root.prev = root;
+    }
+
+    public boolean setExpiration(long expirationMs) {
+        return expiration.getAndSet(expirationMs) != expirationMs;
+    }
+
+    public long getExpiration() {
+        return expiration.get();
+    }
+
+    public synchronized void foreach(Consumer<TimerTask> f) {
+        TimerTaskEntry entry = root.next;
+        while (entry != root) {
+            TimerTaskEntry nextEntry = entry.next;
+            if (!entry.cancelled()) f.accept(entry.timerTask);
+            entry = nextEntry;
+        }
+    }
+
+    public void add(TimerTaskEntry timerTaskEntry) {
+        boolean done = false;
+        while (!done) {
+            // Remove the timer task entry if it is already in any other list
+            // We do this outside of the sync block below to avoid deadlocking.
+            // We may retry until timerTaskEntry.list becomes null.
+            timerTaskEntry.remove();
+
+            synchronized (this) {
+                synchronized (timerTaskEntry) {
+                    if (timerTaskEntry.list == null) {
+                        // put the timer task entry to the end of the list. (root.prev points to the tail entry)
+                        TimerTaskEntry tail = root.prev;
+                        timerTaskEntry.next = root;
+                        timerTaskEntry.prev = tail;
+                        timerTaskEntry.list = this;
+                        tail.next = timerTaskEntry;
+                        root.prev = timerTaskEntry;
+                        taskCounter.incrementAndGet();
+                        done = true;
+                    }
+                }
+            }
+        }
+    }
+
+    public synchronized void remove(TimerTaskEntry timerTaskEntry) {
+        synchronized (timerTaskEntry) {
+            if (timerTaskEntry.list == this) {
+                timerTaskEntry.next.prev = timerTaskEntry.prev;
+                timerTaskEntry.prev.next = timerTaskEntry.next;
+                timerTaskEntry.next = null;
+                timerTaskEntry.prev = null;
+                timerTaskEntry.list = null;
+                taskCounter.decrementAndGet();
+            }
+        }
+    }
+
+    public synchronized void flush(Consumer<TimerTaskEntry> f) {
+        TimerTaskEntry head = root.next;
+        while (head != root) {
+            remove(head);
+            f.accept(head);
+            head = root.next;
+        }
+        expiration.set(-1L);
+    }
+
+    @Override
+    public long getDelay(TimeUnit unit) {
+        return unit.convert(Math.max(getExpiration() - Time.SYSTEM.hiResClockMs(), 0), TimeUnit.MICROSECONDS);
```

Comment:
The timeunit is MILLISECONDS in the scala code here. Is the change to MICROSECONDS intentional? If no, then why didn't a unit test fail? Could we add that missing unit test please?

Reply:
It is not... Let me fix this.